toolchain/base/timings.h

```diff
@@ -0,0 +1,51 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_BASE_TIMINGS_H_
+#define CARBON_TOOLCHAIN_BASE_TIMINGS_H_
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "toolchain/base/yaml.h"
+
+namespace Carbon {
+
+// Helps track timings for a compile.
+class Timings {
+ public:
+  // Adds tracking for nanoseconds, paired with the given label.
+  auto Add(std::string label, std::chrono::nanoseconds nanoseconds) -> void {
+    timings_.push_back({.label = std::move(label), .nanoseconds = nanoseconds});
+  }
+
+  auto OutputYaml(llvm::StringRef filename) const -> Yaml::OutputMapping {
+    // Explicitly copy the filename.
+    return Yaml::OutputMapping([&, filename](Yaml::OutputMapping::Map map) {
+      map.Add("filename", filename);
+      map.Add("nanoseconds",
+              Yaml::OutputMapping([&](Yaml::OutputMapping::Map label_map) {
+                auto total_nanoseconds = std::chrono::nanoseconds(0);
```

Comment:
```suggestion
                std::chrono::nanoseconds total_nanoseconds(0);
```

This is a style preference for direct construction.